ovh/resource_iam_resource_tags.go

```diff
@@ -0,0 +1,217 @@
+
+	keyRegex := regexp.MustCompile(`^[a-zA-Z0-9_.:/=+@-]{1,128}$`)
+	valueRegex := regexp.MustCompile(`^[a-zA-Z0-9_.:/=+@-]{0,256}$`)
+
+	for key, val := range tags {
+		if !keyRegex.MatchString(key) {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Invalid tag key",
+				Detail:   fmt.Sprintf("Tag key '%s' must match pattern ^[a-zA-Z0-9_.:/=+@-]{1,128}$", key),
+			})
+		}
+
+		value := val.(string)
+		if !valueRegex.MatchString(value) {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Invalid tag value",
+				Detail:   fmt.Sprintf("Tag value '%s' for key '%s' must match pattern ^[a-zA-Z0-9_.:/=+@-]{0,256}$", value, key),
+			})
+		}
+	}
+
+	return diags
+}
+
+func resourceIamResourceTagsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Config)
+	resourceURN := d.Get("resource_urn").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	// Track successfully created tags for cleanup on error
+	createdTags := []string{}
+
+	// Create each tag
+	for key, val := range tags {
+		value := val.(string)
+		tagBody := map[string]string{
+			"key":   key,
+			"value": value,
+		}
+
+		endpoint := fmt.Sprintf("/v2/iam/resource/%s/tag", url.PathEscape(resourceURN))
+		err := config.OVHClient.PostWithContext(ctx, endpoint, tagBody, nil)
+		if err != nil {
+			// Best-effort cleanup: delete successfully created tags
+			for _, createdKey := range createdTags {
+				deleteEndpoint := fmt.Sprintf("/v2/iam/resource/%s/tag/%s", url.PathEscape(resourceURN), url.PathEscape(createdKey))
+				_ = config.OVHClient.DeleteWithContext(ctx, deleteEndpoint, nil)
+			}
+			return diag.FromErr(fmt.Errorf("failed to create tag '%s': %w", key, err))
+		}
+		createdTags = append(createdTags, key)
+	}
+
+	// Set the resource ID to the URN
+	d.SetId(resourceURN)
+
+	// Read back the resource to sync state
+	return resourceIamResourceTagsRead(ctx, d, meta)
+}
+
+func resourceIamResourceTagsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Config)
+	resourceURN := d.Id()
+
+	// Fetch the resource details
+	var resourceDetails IamResourceDetails
+	endpoint := fmt.Sprintf("/v2/iam/resource/%s", url.PathEscape(resourceURN))
+	err := config.OVHClient.GetWithContext(ctx, endpoint, &resourceDetails)
+	if err != nil {
+		// If resource not found, remove from state
+		if errOvh, ok := err.(*ovh.APIError); ok && errOvh.Code == 404 {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("failed to read resource: %w", err))
+	}
+
+	// Set resource_urn
+	if err := d.Set("resource_urn", resourceDetails.URN); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Get tags from config
+	configTags := d.Get("tags").(map[string]interface{})
+	managedTags := make(map[string]string)
+
+	// If config is empty (e.g., during import), read all tags from API
+	if len(configTags) == 0 {
```

Comment:
I'm not sure to understand why you need this `if/else` ? From my POV you should only exclude tags that are prefixed by `ovh:` as you cannot change them, but for the rest you could just set them in the state.

Reply:
Are there some resources with `ovh:` prefixed tags that can be created without incurring costs?
I'd like to add acceptance tests to exercise this special case.

So far, the only resource I've identified with such tags are `ovh_ip_service` but they are billed €1.5 per month.